### DISCUSSION TITLE
QR and barcode consistency

### DISCUSSION THREAD

**Discussion by Petrox** [inventree/InvenTree#5689] [General]
I know Qrcode and barcode plugins are not full scale solutions, more like a built-in-demo, so I'm only talking about the features of the demo implementation here:

For stock locations I have two options:
 - Qrcode only
 - Qrcode with label

For Part:
 - Simple Part label (which is actually the Qrcode with label)
 - Barcode part label

So we have 3 different outputs for two classes. One can't print barcodes for stock locations, or they can't print QRcode only for parts.

Then I go to a stockitem:

 - I don't have label printing option under the same icon (just "view qrcode"), but there is a Printer icon now (not seen on Parts or Stocklocations).
 - My only option is "Simple Qr code label", which doesn't contain the name of the product nor the location nor batch number nor serial number.
 - There is no Barcode option here either

It's not complaining, just taking some time to record my experiences with the labels on current stable (0.12.8) release. And I think the already existing features would only need a bit of configuration to offer a consistent interface towards the user.

**Comment by SchrodingersGat** [comment]
@Petrox this really only comes down to the default templates we provide - technically it would be possible of course. This would be a pretty good "getting started" issue for someone to add these other default templates. Would it be something you would be interested in contributing?

**Reply by Petrox** [reply]
I would be happy to contribute, only my spare time is limited. customizing some templates shouldn't be too difficult, so it all depends on the environment setup, so let me take a look. I might have to ask a few basic questions, but we'll see.

**Reply by SchrodingersGat** [reply]
>  I might have to ask a few basic questions, but we'll see.

Happy to provide whatever assistance you need to get up and running :)

**Reply by Petrox** [reply]
> > I might have to ask a few basic questions, but we'll see.
> 
> Happy to provide whatever assistance you need to get up and running :)

Dear @SchrodingersGat, 

I have prepared a set of qrcode label templates according to my own needs, trying to cover:
 - qr codes with names for parts and stockitems (similar to stock locations)
 - stock items with serial numbers and without serialnumbers
 - different lengths (qrcode only, compact = 24x50mm, medium 24x70mm, and long 24x100mm) for the label printers

 One additional trick is to make the PART labels have the QRcode on the right side (while stocklocations and stockitems have them on the left)  reason: stocklocations (boxes) might have both the location labels and the part labels (which are stored there) on them and a subconscious differentiation between them helps the user to scan what he is looking for. 
 There could be reports or specific boxes where PART and STOCKITEM labels are both attached, in that situation the LEFT-RIGHT difference comes handy again.
 Having STOCKLOCATION + STOCKITEMS on the same box is not a valid scenario (in my case).
 
 Are you interested in receiving a pull request that would allow a newcomer to have qrcodes working for parts-stocklocations-stockitems out of the box?

**Reply by Petrox** [reply]
I'm not on the latest version yet though, so I might be late from the party.

**Reply by SchrodingersGat** [reply]
@Petrox happy to receive a PR to improve the "out of the box" templates for sure!

**Comment by BorampetBhavani** [comment]
Hello I am new to this platform and I am a recent graduate student . when i am working on inventree , when i scan any barcode its not giving any result.  app version is 0.13.3 and also how do i need to update the version please help me out.

**Reply by Petrox** [reply]
Dear BorampetBhavani,

This issue is quite different from what you wish to solve, but let me still try to help a bit:

 I have checked the barcodes with 0.12.10 and the latest current stable android app and barcode scanning works perfectly with the default part template and qrcode scanning also works perfectly. Note, both 12.10 and 0.13.3 is are far from the latest versions, but I'm quite confident the barcode printing and scanning didn't change much.

 You have not clarified what kind of barcode reader you've been using, maybe you are using a serial or HID device, what is not capable or not properly configured to read this specific barcode. Please read the manual of the barcode scanner throughly, these devices often need configuration through special barcode sheets of labels.

Should you be able to read the codes with a random mobile app or barcode scanner (and not using the Inventree mobile app configured to connect to your inventree instance) then you will only read a reference to the database object:
 - The barcode and qrcodes only contain the type and id of the objects in the database (eg.: part:97 or stockitem:103), nothing else.
 - This means the barcodes themselves are not useful without knowing which instance of inventree it was generated from. (Or in a different instance it would point to random or invalid parts/stockitems).

 Should you have more questions please open a new specific github issue, this one is a different story.
